protocol/network/wire.md

```diff
@@ -0,0 +1,105 @@
+
+Protocol Multiplexing means running multiple different protocols over the same stream. This could happen sequentially (one after the other), or concurrently (at the same time, with their messages interleaved). We achieve protocol multiplexing using three pieces:
+
+- [multistream](#multistream) - self-describing protocol streams
+- [multistream-select](#multistream-select) - a self-describing protocol selector
+- [Stream Multiplexing](#stream-multiplexing) - running many independent streams over the same wire.
+
+### multistream - self-describing protocol stream
+
+[multistream](https://github.com/jbenet/multistream) is a self-describing protocol stream format. It is extremely simple. Its goal is to define a way to add headers to protocols that describe the protocol itself. It is sort of like adding versions to a protocol, but being extremely explicit.
+
+For example:
+
+```
+/ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/0.2.3
+<dht-message>
+<dht-message>
+...
+```
+
+### multistream-selector - self-describing protocol stream selector
+
+[multistream-select](https://github.com/jbenet/multistream/tree/master/multistream-select) is a simple [multistream](https://github.com/jbenet/multistream) protocol that allows listing and selecting other protocols. This means that Protomux has a list of registered protocols, listens for one, and then _nests_ (or upgrades) the connection to speak the registered protocol. This takes direct advantage of multistream: it enables interleaving multiple protocols, as well as inspecting what protocols might be spoken by the remote endpoint.
+
+For example:
+
+```
+/ipfs/QmdRKVhvzyATs3L6dosSb6w8hKuqfZK2SyPVqcYJ5VLYa2/multistream-select/0.3.0
+/ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/0.2.3
+<dht-message>
+<dht-message>
+...
+```
+
+### Stream Multiplexing
+
+Stream Multiplexing is the process of multiplexing (or combining) many different streams into a single one. This is a complicated subject because it enables protocols to run concurrently over the same wire. And all sorts of notions regarding fairness, flow control, head-of-line blocking, etc. start affecting the protocols. In practice, stream multiplexing is well understood and there are many stream multiplexing protocols. To name a few:
+
+- HTTP/2
+- SPDY
+- QUIC
+- SSH
+
+IPFS nodes are free to support whatever stream multiplexors they wish, on top of the default one. The default one is there to enable even the simplest of nodes to speak multiple protocols at once. The default multiplexor will be HTTP/2 (or maybe QUIC?), but implementations for it are sparse, so we are beginning with SPDY. We simply select which protocol to use with a multistream header.
+
+For example:
+
+```
+/ipfs/QmdRKVhvzyATs3L6dosSb6w8hKuqfZK2SyPVqcYJ5VLYa2/multistream-select/0.3.0
+/ipfs/Qmb4d8ZLuqnnVptqTxwqt3aFqgPYruAbfeksvRV1Ds8Gri/spdy/3
+<spdy-header-opening-a-stream-0>
+/ipfs/QmVXZiejj3sXEmxuQxF2RjmFbEiE9w7T82xDn3uYNuhbFb/ipfs-dht/0.2.3
```

Comment:
i should probably make real examples that are resolvable. the idea is that you'd have the specs in directories that give the protocol name + version number, and then we'd publish them with ipfs, getting a resulting root hash. this allows the hashes to be resolved for and look at the spec. but yeah, _these_ particular hashes aren't resolvable because i made them up. perhaps they should be for more clear examples. (may also be relevant to link directly to commit objects once we have those.)